src/csharp/Grpc.Core/Server.cs

```diff
@@ -316,6 +319,20 @@ private void AllowOneRpc(CompletionQueueSafeHandle cq)
             }
         }
 
+        /// <summary>
+        /// Checks that all ports have been bound successfully.
+        /// </summary>
+        private void CheckPortsBoundSuccessfully()
+        {
+            lock (myLock)
+            {
+                if (!ports.All((port) => port.BoundPort != 0))
+                {
+                    throw new IOException("Failed to bind some of ports exposed by the server.");
```

Comment:
If the unbound ports (or at least one of them) were included in the exception message, that would make it more actionable. Though, I'm not sure whether having custom messages in exceptions is OK in this code base. (Some people like to keep all their exception messages static to help with localization, for example.)

Something like this?

```
var unboundPort = ports.FirstOrDefault(port => port.BoundPort == 0);
if (unboundPort != null)
{
    // may need to pass a culture: don't know this code base's style
    throw new IOException(
        string.Format("Failed to bind port \"{0}:{1}\"", unboundPort.Host, unboundPort.Port));
}
```